contracts/StorageFees.cdc

```diff
@@ -0,0 +1,244 @@
+//
+// An account moving/deleting its `StorageFees.StorageReservation` resource will result in the transaction failing because the account will have no storage capacity.
+// Moving the `StorageFees.StorageReservationReceiver` will result in the account being unable to receive or withdraw reserved flow. This can be fixed by calling: 
+// ```
+// account.link<&StorageFees.StorageReservation{StorageFees.StorageReservationReceiver}>(
+//     StorageFees.storageReservationPath,
+//     target: StorageFees.storageReservationReceiverPath)
+// ```
+pub contract StorageFees {
+    // Emitted when storage capacity refunding is enabled or disabled.
+    pub event RefundingEnabledChanged(_ enabled: Bool)
+    
+    // Emitted when the amount of storage capacity an account has per reserved Flow token changes
+    pub event StorageBytesPerReservedFlowChanged(_ storageBytesPerReservedFlow: UFix64)
+
+    // Emitted when minimum amount of Flow 
+    pub event MinimumStorageReservationChanged(_ minimumStorageReservation: UFix64)
+
+    // Emitted when the minimum amount of Flow tokens that an account needs to have reserved for storage capacity changes.
+    pub event StorageReservationChanged(address: Address, oldStorageReservation: UFix64, oldStorageCapacity: UInt64, newStorageReservation: UFix64, newStorageCapacity: UInt64)
+
+    // Defines the path where each account should have a `StorageReservationReceiver` capability
+    pub let storageReservationReceiverPath: PublicPath
+
+    // Defines the path where each account should have a `StorageReservation` capability
+    pub let storageReservationPath: StoragePath
+
+    // Defines how much storage capacity an account has per reserved Flow token.
+    // definition is written per unit of flow instead of the inverse, so there is no loss of precision calculating storage from flow, but there is loss of precision when calculating flow per storage.
+    pub var storageBytesPerReservedFlow: UFix64
+
+    // Defines the minimum amount of Flow tokens that an account needs to have reserved for storage capacity.
+    // If an account has less then this amount reserved by the end of any transaction it participated in, the transaction will fail.
+    pub var minimumStorageReservation: UFix64
+
+    // Enables or disables the refunding storage capacity.
+    pub var refundingEnabled: Bool
+
+    // An administrator resource that can change the parameters of the StorageFees smart contract.
+    pub resource Administrator {
+        pub fun setRefundingEnabled(_ enabled: Bool) {
+            if StorageFees.refundingEnabled == enabled {
+              return
+            }
+            StorageFees.refundingEnabled = enabled
+            emit RefundingEnabledChanged(enabled)
+        }
+
+        // Changes the amount storage capacity an account has per accounts' reserved storage.
+        pub fun setStorageBytesPerReservedFlow(_ storageBytesPerReservedFlow: UFix64) {
```

Comment:
Are there any pre-conditions on this that could check that the argument is not a unexpected value? I'm not sure if there are any disallowed values

Reply:
I'm not sure....
the only potential problem would be setting it to 0.0 but that transaction would definitely fail as the service account would not have sufficient storage anymore.